### PR TITLE
DM-31964: Fixup dimensions on ProcessCcdWithFakesTask/MatchFakesTask

### PR DESCRIPTION
**Pull request by morriscb**
Add code to concatenate fake catalogs.

Fix linting.

Add compose fakes to variablity code.

Debug matchFakes.

Add docs

Simplify unittests.

Remove unnessecary runQuantum test.

Change coding dimensions.

Fix linting.

Output tract numbers.

Fix linting.

Access tractId

Access tractId

Fix ra/decColName usage.

### CODE REVIEW

File: python/lsst/pipe/tasks/matchFakes.py [hunk 1]
```diff
@@ -148,6 +178,41 @@ def run(self, fakeCat, diffIm, associatedDiaSources):
...
+        """
+        if len(fakeCats) == 1:
+            return fakeCats[0].get(
+                datasetType=self.config.connections.fakeCats,
+                immediate=True)
+        outputCat = []
+        for fakeCatRef in fakeCats:
+            cat = fakeCatRef.get(
+                datasetType=self.config.connections.fakeCats,
+                immediate=True)
```
Comment (timj) [inline python/lsst/pipe/tasks/matchFakes.py:205]: There is no `immediate` keyword for `get()`.

File: python/lsst/pipe/tasks/matchFakes.py [hunk 2]
```diff
@@ -148,6 +178,39 @@ def run(self, fakeCat, diffIm, associatedDiaSources):
...
+                datasetType=self.config.connections.fakeCats)
+        outputCat = []
+        for fakeCatRef in fakeCats:
+            cat = fakeCatRef.get(
+                datasetType=self.config.connections.fakeCats)
+            tractId = fakeCatRef.dataId["tract"]
+            # Make sure all data is within the inner part of the tract.
+            outputCat.append(cat[
+                skyMap.findTractIdArray(cat[self.config.raColName],
+                                        cat[self.config.decColName],
```
Comment (surhudm) [inline python/lsst/pipe/tasks/matchFakes.py:208]: @morriscb shouldn't these column names be `self.config.ra_col` and `self.config.dec_col` given that `self.config.raColName` and `self.config.decColName` are going to be deprecated?

File: python/lsst/pipe/tasks/processCcdWithFakes.py [hunk 3]
```diff
@@ -56,11 +64,15 @@ class ProcessCcdWithFakesConnections(PipelineTaskConnections,
         dimensions=("instrument", "visit", "detector")
     )
 
-    fakeCat = cT.Input(
-        doc="Catalog of fake sources to draw inputs from.",
+    fakeCats = cT.Input(
+        doc="Set of catalogs of fake sources to draw inputs from. We will "
+            "from one of these that fully contains our exposure or the is "
+            "closest the visit boresight.",
```
Comment (surhudm) [inline python/lsst/pipe/tasks/processCcdWithFakes.py:70]: @morriscb I guess the description here has a word missing between L68 and L69. L69 has `the is`. I believe the description also does not reflect the fact that it gathers fake sources from other neighboring tracts as done in `composeFakeCat`.

File: python/lsst/pipe/tasks/processCcdWithFakes.py [hunk 4]
```diff
@@ -445,6 +474,39 @@ def run(self, fakeCat, exposure, wcs=None, photoCalib=None, exposureIdInfo=None,
...
+        outputCat = []
+        for fakeCatRef in fakeCats:
+            cat = fakeCatRef.get(
+                datasetType=self.config.connections.fakeCats)
+            tractId = fakeCatRef.dataId["tract"]
+            # Make sure all data is within the inner part of the tract.
+            outputCat.append(cat[
+                skyMap.findTractIdArray(cat[self.config.insertFakes.raColName],
+                                        cat[self.config.insertFakes.decColName],
+                                        degrees=False)
```
Comment (surhudm) [inline python/lsst/pipe/tasks/processCcdWithFakes.py:505]: @morriscb Same comment as above: `self.config.insertfakes.raColName` and `self.config.insertFakes.decColName` should be changed to `self.config.ra_col` and `self.config.dec_col` to remove deprecated column names.

File: python/lsst/pipe/tasks/processCcdWithFakes.py [hunk 5]
```diff
@@ -343,29 +359,38 @@ def runQuantum(self, butlerQC, inputRefs, outputRefs):
...
             inputs["wcs"] = row.getWcs()
-
         elif self.config.doApplyExternalTractSkyWcs:
-            externalSkyWcsCatalog = inputs["externalSkyWcsTractCatalog"]
+            externalSkyWcsCatalogList = inputs["externalSkyWcsTractCatalog"]
+            for externalSkyWcsCatalogRef in externalSkyWcsCatalogList:
+                if externalSkyWcsCatalogRef.dataId["tract"] == tractId:
+                    externalSkyWcsCatalog = externalSkyWcsCatalogRef.get(
+                        datasetType=self.config.connections.externalSkyWcsTractCatalog)
+                    break
```
Comment (surhudm) [inline python/lsst/pipe/tasks/processCcdWithFakes.py:377]: @morriscb I have seen this return in some places with no `externalSkyWcsCatalog` because the condition is never satisfied. You could let it use the last `externalSkyWcsCatalog` with an else statement like:

```
else:
    externalSkyWcsCatalog = externalSkyWcsCatalogRef.get(datasetType=self.config.connections.externalSkyWcsTractCatalog)
```

Similar problem for the photoCalib below in addition to a bug (see below L391).

File: python/lsst/pipe/tasks/processCcdWithFakes.py [hunk 6]
```diff
@@ -343,29 +359,38 @@ def runQuantum(self, butlerQC, inputRefs, outputRefs):
...
             row = externalPhotoCalibCatalog.find(detectorId)
             inputs["photoCalib"] = row.getPhotoCalib()
-
         elif self.config.doApplyExternalTractPhotoCalib:
-            externalPhotoCalibCatalog = inputs["externalPhotoCalibTractCatalog"]
+            externalPhotoCalibCatalogList = inputs["externalPhotoCalibTractCatalog"]
+            for externalPhotoCalibCatalogRef in externalPhotoCalibCatalogList:
+                if externalPhotoCalibCatalogRef.dataId["tract"] == tractId:
+                    externalPhotoCalibCatalog = externalPhotoCalibCatalogRef.get(
+                        datasetType=self.config.connections.externalSkyWcsTractCatalog)
```
Comment (surhudm) [inline python/lsst/pipe/tasks/processCcdWithFakes.py:392]: @morriscb Shouldn't this be the externalPhotoCalibCatalog connection instead of the SkyWcsTractCatalog? @sr525 spotted this.

File: python/lsst/pipe/tasks/matchFakes.py [hunk 7]
```diff
@@ -157,10 +220,12 @@ def _trimFakeCat(self, fakeCat, image):
             The catalog of fake sources to be input
         image : `lsst.afw.image.exposure.exposure.ExposureF`
             The image into which the fake sources should be added
+        skyMap : `lsst.skymap.SkyMap`
+            SkyMap defining the tracts and patches the fakes are stored over.
 
         Returns
         -------
-        fakeCat : `pandas.core.frame.DataFrame`
+        fakeCats : `pandas.core.frame.DataFrame`
```
Comment (leeskelvin) [inline python/lsst/pipe/tasks/matchFakes.py:228]: Please explain why this has been pluralized in the description for the parameter.

File: python/lsst/pipe/tasks/processCcdWithFakes.py (hunk 3 above)
Comment (leeskelvin) [reply python/lsst/pipe/tasks/processCcdWithFakes.py:70]: As above, the doc string here doesn't read well - can you rephrase?

File: python/lsst/pipe/tasks/processCcdWithFakes.py (hunk 5 above)
Comment (leeskelvin) [reply python/lsst/pipe/tasks/processCcdWithFakes.py:377]: +1 for @surhudm's thinking above - it would be good for this large if/elif block to have a fall-back option in place.

File: python/lsst/pipe/tasks/processCcdWithFakes.py (hunk 6 above)
Comment (leeskelvin) [reply python/lsst/pipe/tasks/processCcdWithFakes.py:392]: +1 again - seems like a copy/paste typo.

File: tests/test_matchFakes.py [hunk 8]
```diff
@@ -197,7 +124,8 @@ def testTrimCat(self):
         """Test that the correct number of sources are in the ccd area.
         """
         matchTask = MatchFakesTask()
-        result = matchTask._trimFakeCat(self.fakeCat, self.exposure)
+        result = matchTask._trimFakeCat(
+            self.fakeCat, self.exposure)
```
Comment (leeskelvin) [inline tests/test_matchFakes.py:128]: This doesn't need to be on a new line.

File: python/lsst/pipe/tasks/processCcdWithFakes.py (hunk 5 above)
Comment (morriscb) [reply python/lsst/pipe/tasks/processCcdWithFakes.py:377]: If the condition is never satisfied, that means there is no wcs or photocalib for that tract. I'm not sure what the safest fall back position. Likely people doing who have written the analysis code and use these calibs should decide.
